File: common/common-persistence-jooq/src/main/java/ch/difty/scipamato/common/persistence/AbstractFilterConditionMapper.java
package ch.difty.scipamato.common.persistence;

import java.util.ArrayList;
import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jooq.Condition;
import org.jooq.impl.DSL;

import ch.difty.scipamato.common.entity.filter.ScipamatoFilter;

/**
 * Abstract class providing the basic algorithm as template method to map the
 * filter into the conditions.
 *
 * @param <F>
 *     Type of the filter class, extending {@link ScipamatoFilter}
 * @author u.joss
 */
public abstract class AbstractFilterConditionMapper<F extends ScipamatoFilter>
    implements GenericFilterConditionMapper<F> {

    @NotNull
    @Override
    public final Condition map(@Nullable final F filter) {
        final List<Condition> conditions = new ArrayList<>();
        if (filter != null)
            map(filter, conditions);
        return DSL.and(conditions);
    }

    /**
     * Maps the actual filter attributes into the provided list of conditions.
     *
     * @param filter
     *     provides the filter attributes to be mapped
     * @param conditions
     *     receives of the {@link Condition}s derived from the filter attributes
     */
    protected abstract void map(@NotNull F filter, @NotNull List<Condition> conditions);

}

File: common/common-persistence-jooq/src/main/kotlin/ch/difty/scipamato/common/persistence/AbstractFilterConditionMapper.kt
package ch.difty.scipamato.common.persistence

import ch.difty.scipamato.common.entity.filter.ScipamatoFilter
import org.jooq.Condition
import org.jooq.impl.DSL

/**
 * Abstract class providing the basic algorithm as template method to map the
 * filter into the conditions.
 *
 * @param <F> Type of the filter class, extending [ScipamatoFilter]
 */
abstract class AbstractFilterConditionMapper<in F : ScipamatoFilter> : GenericFilterConditionMapper<F> {

    final override fun map(filter: F?): Condition {
        val conditions = filter?.let { internalMap(it) } ?: emptyList()
        return DSL.and(conditions)
    }

    /**
     * Maps the actual [filter] attributes into the provided list of [Condition]s.
     */
    protected abstract fun internalMap(filter: F): List<Condition>
}
